### DISCUSSION TITLE
Still completion

### DISCUSSION THREAD

**Discussion by SoilihBenSoilih** [c-cube/linol#11] [General]
Hello,

I hope you are well !
I am writing to you because I need help with completion. I am using Linol to develop an lsp server, I am having trouble getting the completion to work. I made a little code that should work statically with simple data for some tries, but the completion doesn't work, I don't know the reason, can you take a look if it Please ?

[linol_lsp.txt](https://github.com/c-cube/linol/files/6474269/linol_lsp.txt)

Also, how do I specify that, for example, if I type `List.` the completion which follows is only the methods of the `List module` for example?

Thank you very much, best to you !!!

**Comment by c-cube** [comment]
Je ne vois rien de mauvais à première vue, donc aucune idée. Essayez de logger ce qui se passe (et de vérifier qu'au moins le document est mis à jour proprement, etc.)

Pour lister dans `List.<tab>` ça dépend du langage en question, et c'est en dehors du scope de Linol pour être honnête. Il faut voir ce que que le compilateur donne comme interface.

**Reply by SoilihBenSoilih** [reply]
Oooh, tu parles français, superbe !!!
En fait, j'utilise emacs, j'ai regardé dans les logs les requêtes envoyées et les réponses reçues,
et il n'y a pas de requêtes de complétion envoyée, donc évidemment, on a pas de réponses, et c'est ça
qui fait que je n'ai pas de complétion.

**Reply by SoilihBenSoilih** [reply]
Hello, 
I hope you are doing well !
I've found the problem but i don't know how to solve that so I was wondering if you could help me please.
I have taken a look on the initialization request. 
And the problem is that, the initialization doesn't have completionProvider, and, for some clients, when there is 
not completionProvider, so there is not completion.
Here is the log of the initialization request.
[initialize.txt](https://github.com/c-cube/linol/files/6489373/initialize.txt)

And here a basic server with just static diagnotics and and completions
to test if it works.

[linol_lsp.txt](https://github.com/c-cube/linol/files/6489379/linol_lsp.txt)

Can you, please, help me to solve that?

Thank you again a lot, best to you !!!

**Reply by c-cube** [reply]
Seems like I missed that one in the `config_` methods. You might be able to use `config_modify_capabilities` to set the completion provider properly.

**Reply by SoilihBenSoilih** [reply]
Oh well !
I have written something like that : 

```
method! config_modify_capabilities compl = 
      let t = Lsp.Types.CompletionOptions.create () in 
      Lsp.Types.ServerCapabilities.create ~completionProvider:t ()
```
But it still not working... :-(
Isn't that the right way?

It show me completion Provider, but nothing inside in the log : 

```
Result: {
  "capabilities": {
    "completionProvider": {
    }
  }
}
```

**Reply by c-cube** [reply]
idk, you'll have to dig deeper on what emacs does. You might have to give explicit options to `CompletionOptions.create`.

**Reply by SoilihBenSoilih** [reply]
Hello, 
I hope you're doing well.
I have done a lot of things in the client side for completion
and changed the initialization. I've compared it to the initialiation
of the TypeScript Lsp server ( 
[diff.pdf](https://github.com/c-cube/linol/files/6509084/diff.pdf) ), and 
something missed, the triggged, so i added them and now, the request 
is sent, but it's return `null`, so the problem is from my server, but i don't
know what i did wrong, I have done a static server with just completion and
diagnostics which is here : 
[linol_lsp.txt](https://github.com/c-cube/linol/files/6516423/linol_lsp.txt)

I have another problem. when I use the function `config_modify_capabilities`
to add completionProvider in the initial request, the result of the initial request
change and become : 

```
Result: {
  "capabilities": {
    "completionProvider": {
      "resolveProvider": true,
      "triggerCharacters": [
        ".",
        "#",
        "\"",
        "'",
        "/",
        "@",
        "<"
      ]
    }
  }
}
```
And before, without overriden method `config_modify_capabilities`, it was
this : 

```
Result: {
  "capabilities": {
    "executeCommandProvider": {
      "commands": []
    },
    "codeActionProvider": null,
    "definitionProvider": true,
    "hoverProvider": true,
    "textDocumentSync": {
      "willSave": null,
      "change": 2
    }
  }
}
```

So, after solving the completionProvider, hover is not supported no more, because 
it don't appear in the results no more like if completionProvider have overriden the initialization.

I have these two problems to fix. I need your help please.

Thank you very much !

**Reply by c-cube** [reply]
First: I can't be the only one helping you, can I? Is there no one else that knows OCaml around you?

Second: seems like in `config_modify_capabilities` you didn't modify the input capabilities, but returned a fresh one instead. Am I wrong?

**Reply by SoilihBenSoilih** [reply]
Hello,
I'm sorry to ask you so much. Actually, I'm still a student and this is my graduation project, I've never worked with ocaml before, but noticeably only with object-oriented languages, and had to start directly with an LSP server, this which in itself is not something simple for someone who has never programmed in ocaml. And it is because it is you who made Linol, by the way, I thank you very much, that it is to you that I ask questions because I do not know many people who happen to me help move forward. I'm sorry to ask you so many questions. Also, if I didn't change the input capabilities, it's because I didn't know how to do it. I'm sorry ...

**Reply by c-cube** [reply]
Yeah no sweat. You may want to join the OCaml discord or #ocaml on libera.chat to get help from more people.

if the definition is public, it might look like (roughly):

```ocaml
method config_modify_capabilities caps =
  let foo = … in
  { caps with Lsp.Types.Capabilities.completionProvider = foo }
```

which returns a new record similar to `caps` but with the fields `completionProvider` set to `foo`.

**Reply by SoilihBenSoilih** [reply]
Ooh, ok thank you !
I got it now. Do you have the link of the discord channel or something? When i search for `ocaml` in discord, i don't find nothing.
Thank you again for you time.

**Reply by c-cube** [reply]
https://discord.gg/Jwy5HpVm

**Reply by SoilihBenSoilih** [reply]
Thank you very much !
